### PR TITLE
[ filesystem backend] Add dcrtime database fsck

### PR DESCRIPTION
**Pull request by marcopeereboom**
This code goes through all timestamp directories and verifies that the contents match the global database and the flush record. It identified some results of bugs that were fixed during the duration of the live version.

### CODE REVIEW

File: cmd/dcrtime_fsck/dcrtime_fsck.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	fs, err := filesystem.NewDump(root)
+	if err != nil {
+		return err
+	}
+	defer fs.Close()
+
+	return fs.Fsck(&backend.FsckOptions{
+		Verbose:     *verbose,
+		PrintHashes: *printHashes,
+		Fix:         *fix,
```
Comment (lukebp) [inline cmd/dcrtime_fsck/dcrtime_fsck.go]: Fix is never actually referenced in the fsck code.  Errors get fixed regardless of whether this is true or false.

File: dcrtimed/backend/filesystem/dump.go [hunk 2]
```diff
@@ -370,3 +376,479 @@ func (fs *FileSystem) Restore(f *os.File, verbose bool, location string) error {
...
+			return fmt.Errorf("   *** ERROR merkle root not "+
+				"found: tx %v merkle %x", flushRecord.Tx,
+				flushRecord.Root)
+		}
+
+		// We are done
+		return nil
+	}
+
+	// 3. Make sure timestamps do not exist in the global database
```
Comment (lukebp) [inline dcrtimed/backend/filesystem/dump.go]: Isn't this `2.` according to the function declaration comment, not `3.`?

File: dcrtimed/backend/filesystem/dump.go [hunk 3]
```diff
@@ -370,3 +376,479 @@ func (fs *FileSystem) Restore(f *os.File, verbose bool, location string) error {
...
+		gdbts, err := fs.db.Get(key, nil)
+		if err != nil {
+			if err == leveldb.ErrNotFound {
+				continue
+			}
+			return fmt.Errorf("   *** ERROR found in db: %v %v",
+				k, err)
+		}
+		dbts := int64(binary.LittleEndian.Uint64(gdbts))
+		if dbts != ts {
```
Comment (lukebp) [inline dcrtimed/backend/filesystem/dump.go]: What about the scenario where `dbts == ts`?  

`dbts == ts` would mean that the global db is pointing to the correct timestamp for that digest, but the timestamp is missing a flush records so it shouldn't be in the global db in the first place.  The global db record should be deleted.

File: dcrtimed/backend/filesystem/dump.go [hunk 4]
```diff
@@ -370,3 +376,479 @@ func (fs *FileSystem) Restore(f *os.File, verbose bool, location string) error {
...
+			fmt.Printf("    *** ERROR duplicate key: %v %v %v\n",
+				ts, v, k)
+
+			// 1. Verify against global db
+			ok, err := fs.db.Has(key, nil)
+			if err != nil {
+				return fmt.Errorf("   *** ERROR duplicate "+
+					"key: has %v", err)
+			}
+			if ok {
```
Comment (lukebp) [inline dcrtimed/backend/filesystem/dump.go]: Won't `ok` always be true since we already know that the digest is a duplicate?  

Don't we need to make sure that we're deleting the correct duplicate by retrieving the timestamp from the global database and checking if it matches `ts`.  If it doesn't match then we delete the digest from the current timestamp db.

File: dcrtimed/backend/filesystem/dump.go (hunk 2 above)
Comment (marcopeereboom) [reply dcrtimed/backend/filesystem/dump.go]: I flipped the code around and that got massacred. Fixed.

File: cmd/dcrtime_fsck/dcrtime_fsck.go (hunk 1 above)
Comment (marcopeereboom) [reply cmd/dcrtime_fsck/dcrtime_fsck.go]: Also got lost somehow. Adding it back.

File: dcrtimed/backend/filesystem/dump.go (hunk 4 above)
Comment (marcopeereboom) [reply dcrtimed/backend/filesystem/dump.go]: Actually the other way around. The test is in there to make sure we don't catch false-positives.

So the dup list ONLY is filled when there is no flush record AND when there are multiple entries for this hash. This means that since it has never been flushed it does not matter which one is deleted and one could argue that the first one IS the correct one since the follow up ones were newer submissions due to the bug described in the comments. I think this is working correctly.

File: dcrtimed/backend/filesystem/dump.go (hunk 3 above)
Comment (marcopeereboom) [reply dcrtimed/backend/filesystem/dump.go]: I am not following this. The subtlety here is that way early in the function we exit if there is no flush record, therefore, we can't get into the condition you describe. Am I missing something?

Comment (lukebp) [reply dcrtimed/backend/filesystem/dump.go]: Where is the exit for when `len(digests) > 0 && flushRecord == nil`?  I'm not seeing it.

Comment (marcopeereboom) [reply dcrtimed/backend/filesystem/dump.go]: That is in order to deal with a timestamp that was completely emptied after fixes around line 575. So a timestamp had let's say one entry and it was deleted as a timestamp mismatch then the map entry is deleted too. If we don't then we'd have to run fsck twice to clean up after that. It is a bit of hack to prevent that. Should I comment on it?

File: dcrtimed/backend/filesystem/dump.go (hunk 4 above)
Comment (lukebp) [reply dcrtimed/backend/filesystem/dump.go]: > So the dup list ONLY is filled when there is no flush

Maybe I'm missing something, but I'm not seeing this in the code.

File: cmd/dcrtime_fsck/dcrtime_fsck.go (hunk 1 above)
Comment (lukebp) [reply cmd/dcrtime_fsck/dcrtime_fsck.go]: Resolved.

File: dcrtimed/backend/filesystem/dump.go (hunk 2 above)
Comment (lukebp) [reply dcrtimed/backend/filesystem/dump.go]: Resolved.

File: dcrtimed/backend/filesystem/dump.go (hunk 4 above)
Comment (marcopeereboom) [reply dcrtimed/backend/filesystem/dump.go]: The dups map is only filled when there are NO flush records. Look at the code around line 750. There is a continue if there is a flush record.

Comment (marcopeereboom) [reply dcrtimed/backend/filesystem/dump.go]: Sorry, that is incorrect. Let me rephrase.

This code iterates over all timestamps and adds all digests to a map. As it is building the map it looks for dups (which is an invalid condition). When it finds a dup it ensures it doesn't exist in the global database (meaning there is no flush record!). I added code that verifies if it does exist in the database that it errors out. Also added code to print all hashes and stuff during phase 3 (when verbose and printhashes) so that these claims can be cross verified.

File: cmd/dcrtime_fsck/dcrtime_fsck.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/decred/dcrtime/dcrtimed/backend"
+	"github.com/decred/dcrtime/dcrtimed/backend/filesystem"
+)
+
+var (
+	defaultHomeDir = dcrutil.AppDataDir("dcrtimed", false)
+
+	file        = flag.String("file", "", "journal of modifications if used (will be written despite -fix)")
+	fix         = flag.Bool("fix", false, "Try to correct correctable failures")
+	dcrtimeHost = flag.String("host", "", "dcrtime block explorer")
```
Comment (dajohi) [inline cmd/dcrtime_fsck/dcrtime_fsck.go]: rename to something better, such as dcrBlockExplorer perhaps.

Comment (jrick) [reply cmd/dcrtime_fsck/dcrtime_fsck.go]: call it dcrdataHost since it uses the dcrdata api (and would not work with an insight block explorer)

File: cmd/dcrtime_dumpdb/dcrtime_dumpdb.go [hunk 6]
```diff
@@ -41,12 +41,15 @@ func _main() error {
 
 	// Dump
 	var root string
+	// XXX this needs to come out of chaincfg
 	if *testnet {
-		root = filepath.Join(*fsRoot, "data", "testnet2")
+		root = filepath.Join(*fsRoot, "data", "testnet3")
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply cmd/dcrtime_dumpdb/dcrtime_dumpdb.go]: You looked at an old version. I replaced all these with `chaincfg.TestNet3Params.Name`
